### PR TITLE
Inform caller when rocksdb is stalling writes

### PR DESCRIPTION
**Pull request by schischi**
Add a new function in Listener to let the caller know when rocksdb
is stalling writes.

### CODE REVIEW

**Comment by ajkr** [comment]
You can use the "Import to Phabricator" button to trigger the internal CI test runs.

File: include/rocksdb/listener.h [hunk 1]
```diff
@@ -86,6 +86,23 @@ enum class BackgroundErrorReason {
   kMemTable,
 };
 
+enum class WriteStallCondition {
+  kNormal,
+  kDelayed,
+  kCompactionPressure,
```
Comment (ajkr) [inline include/rocksdb/listener.h]: can remove this - we don't need to notify for compaction pressure. it's only for an internal mechanism to adjust number of compaction threads.

File: db/db_impl.h [hunk 2]
```diff
@@ -656,8 +656,11 @@ class DBImpl : public DB {
     autovector<MemTable*> memtables_to_free_;
 
     ~WriteContext() {
-      for (auto& sv : superversions_to_free_) {
-        delete sv;
+      for (auto sv : superversions_to_free_) {
+        if (sv != nullptr) {
+          sv->NotifyOnStallConditionsChanged();
```
Comment (ajkr) [inline db/db_impl.h]: unfortunately, I think this approach can cause missed notifications. A compaction or flush can finish with `superversions_to_free_` empty because of a user iterator referencing the same `SuperVersion`, and users can hold their iterators indefinitely.

**Comment by ajkr** [comment]
Were there any issues with storing the stall info directly in JobContext?

**Comment by schischi** [comment]
Some more context:
- The callback is now being called in the job_context
- I still need to store the current stall condition in the SuperVersion because I need access to not only the current condition but also the condition of the previous SuperVersion.
- I couldn't just add the JobContext parameter because, the caller may use a JobContext or a WriteContext. I had to refactor JobContext and WriteContext a little bit.

File: db/db_impl.h [hunk 3]
```diff
@@ -652,13 +652,14 @@ class DBImpl : public DB {
   struct CompactionState;
 
   struct WriteContext {
-    autovector<SuperVersion*> superversions_to_free_;
+    SuperVersionContext superversion_context;
     autovector<MemTable*> memtables_to_free_;
 
+    explicit WriteContext(bool create_superversion = false)
```
Comment (ajkr) [inline db/db_impl.h]: I guess you don't need the argument?

File: db/job_context.h [hunk 4]
```diff
@@ -83,45 +143,36 @@ struct JobContext {
...
     pending_manifest_file_number = 0;
     log_number = 0;
     prev_log_number = 0;
-    new_superversion = create_superversion ? new SuperVersion() : nullptr;
   }
 
   // For non-empty JobContext Clean() has to be called at least once before
   // before destruction (see asserts in ~JobContext()). Should be called with
   // unlocked DB mutex. Destructor doesn't call Clean() to avoid accidentally
   // doing potentially slow Clean() with locked DB mutex.
```
[the remark this answers is not in the document]
Comment (ajkr) [reply db/job_context.h]: I mean in a future refactoring :)

File: db/compacted_db_impl.cc [hunk 5]
```diff
@@ -93,16 +93,18 @@ std::vector<Status> CompactedDBImpl::MultiGet(const ReadOptions& options,
 }
 
 Status CompactedDBImpl::Init(const Options& options) {
+  SuperVersionContext sv_context(true);
```
[the remark this answers is not in the document]
Comment (schischi) [reply db/compacted_db_impl.cc]: Done

File: db/db_impl_compaction_flush.cc [hunk 6]
```diff
@@ -1895,8 +1889,10 @@ SuperVersion* DBImpl::InstallSuperVersionAndScheduleWork(
                         old_sv->mutable_cf_options.max_write_buffer_number;
   }
 
-  auto* old = cfd->InstallSuperVersion(
-      new_sv ? new_sv : new SuperVersion(), &mutex_, mutable_cf_options);
+  if (sv_context->new_superversion == nullptr)
```
[the remark this answers is not in the document]
Comment (schischi) [reply db/db_impl_compaction_flush.cc]: Unfortunately no, in some rare cases the same sv_context may be used twice. 
In this case, only the first call to InstallSuperVersion() gets the pre-allocated pointer, the second call must allocate its own superversion inside of the db_mutex.

See the comment above the 'DBImpl::InstallSuperVersionAndScheduleWork()' for more information.

Also, by default, WriterContext constructs its sv_context with create_superversion = false.
See the message below for more information.

File: db/job_context.h [hunk 7]
```diff
@@ -13,16 +13,78 @@
...
+
+struct SuperVersionContext {
+  struct WriteStallNotification {
+    WriteStallInfo write_stall_info;
+    const ImmutableCFOptions* immutable_cf_options;
+  };
+
+  autovector<SuperVersion*> superversions_to_free;
+  autovector<WriteStallNotification> write_stall_notifications;
+  SuperVersion* new_superversion;  // if nullptr no new superversion
```
[the remark this answers is not in the document]
Comment (schischi) [reply db/job_context.h]: Done.

File: db/job_context.h [hunk 8]
```diff
@@ -13,16 +13,78 @@
...
+    }
+    superversions_to_free.clear();
+    // if new_superversion was not used, it will be non-nullptr and needs
+    // to be freed here
+    delete new_superversion;
+    new_superversion = nullptr;
+  }
+
+  ~SuperVersionContext() {
+    assert(new_superversion == nullptr);
```
[the remark this answers is not in the document]
Comment (schischi) [reply db/job_context.h]: Done

File: db/job_context.h [hunk 9]
```diff
@@ -13,16 +13,78 @@
...
+    assert(superversions_to_free.size() == 0);
+  }
+};
 
 struct JobContext {
   inline bool HaveSomethingToDelete() const {
     return full_scan_candidate_files.size() || sst_delete_files.size() ||
            log_delete_files.size() || manifest_delete_files.size() ||
-           new_superversion != nullptr || superversions_to_free.size() > 0 ||
+           superversion_context.new_superversion != nullptr ||
```
[the remark this answers is not in the document]
Comment (schischi) [reply db/job_context.h]: Nice catch! Fixed.

File: db/job_context.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (schischi) [reply db/job_context.h]: I refactored the code so that all the SuperVersions are always deleted outside of the db_mutex using a sv_context.
So yes, it should be ok to call Clean() from the destructor, but I still agree with the comment: the cleanup is a heavy operation and it should be explicit and not 'hidden' in the destructor.

File: db/db_impl.h (hunk 3 above)
Comment (schischi) [reply db/db_impl.h]: We don't need it, but it may be useful for later optimizations.

Because the implementation of JobContext and WriteContext did not use the same code to deal with SVs, only JobContext was able to pre-allocate its SV, even if it may be useful for WriteContext as well.

I decided to disable it by default, but we might want to enable pre-allocation for some WriteContexts (maybe all of them). This is a bit outside of the scope of this PR, that's why I did not made any changes there, but if you think that it has its place here, I can investigate this.

We can't just blindly enable pre-allocation for all WriteContext, because I'm not sure that all the WriteContext are being used to install SVs and that would be a (small) waste of resource/time.

To sum up, a future improvement would be to check, for each WriteContext, if it is worth pre-allocating the SV or not.

File: db/db_impl.h [hunk 10]
```diff
@@ -1221,15 +1222,11 @@ class DBImpl : public DB {
...
-      ColumnFamilyData* cfd, JobContext* job_context,
-      const MutableCFOptions& mutable_cf_options);
-
   // All ColumnFamily state changes go through this function. Here we analyze
   // the new state and we schedule background work if we detect that the new
   // state needs flush or compaction.
-  SuperVersion* InstallSuperVersionAndScheduleWork(
-      ColumnFamilyData* cfd, SuperVersion* new_sv,
+  void InstallSuperVersionAndScheduleWork(
+      ColumnFamilyData* cfd, SuperVersionContext* job_context,
```
[the remark this answers is not in the document]
Comment (schischi) [reply db/db_impl.h]: Done

File: db/db_impl_compaction_flush.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (ajkr) [reply db/db_impl_compaction_flush.cc]: makes sense, thanks

File: db/db_impl.h (hunk 3 above)
Comment (ajkr) [reply db/db_impl.h]: sounds good

**Comment by ajkr** [comment]
you should still click land, though, else the bot will land this PR under my name, which seems unfair :p
